### ISSUE TITLE
[ENH] plotting constants: implement automatic generation of minimal driver-team mappings for future seasons

### ISSUE THREAD

**Issue by theOehrly** [theOehrly/Fast-F1#619]
### Proposed new feature or change:

Right now, the new plotting module only supports seasons 2018 to 2024 for which team constants are explicitly defined. To improve compatibility of older FastF1 versions with newer seasons, it would be beneficial to implement functionality for generating team constants and driver-team mappings from API data. This could only provide a limited set of information, but it could at least allow the plotting functionality to be useable with reduced functionality.

**Comment by Casper-Guo** [comment]
It would be nice to also get rid of the manual maintenance we had to do for reserve drivers in Q1 when we work on this. So this might need to be an action that runs weekly and updates some data files.

Team constants often require a bit of manual tweaking (for the Fastf1 scheme at least) but only need to be done once a season. So there is less need for automation there?

Writing to data to Python files is not so nice so we should probably move the constants into an actual data configuration format when we do this? Either JSON or TOML would work well.

**Comment by theOehrly** [comment]
> [...] this might need to be an action that runs weekly and updates some data files.

My idea was to implement this directly in FastF1. Updating the code base requires people to update the package, and they will forget. Doing it server-side somewhere with some API is just a huge amount of maintenance incoming.


> Writing to data to Python files is not so nice so we should probably move the constants into an actual data configuration format when we do this? Either JSON or TOML would work well.

Absolutely, the implementation is kind of horrible there, especially the access through globals. I just really wanted to get this done finally when I wrote that, it was a working solution, and it can be changed to something much nicer at any time given that it's only internal stuff.

> It would be nice to also get rid of the manual maintenance we had to do for reserve drivers in Q1 when we work on this.

What exactly do you mean with that? What manual maintenance is necessary there?

**Comment by Casper-Guo** [comment]
> What exactly do you mean with that? What manual maintenance is necessary there?

With the old API, every time a new driver took part in FP1 we had to manually input his abbreviation, team affiliation, and color. My understanding of the current API is we will still have to change the constants to map the drivers to teams. Making that automated is in scope for this issue.

> My idea was to implement this directly in FastF1. Updating the code base requires people to update the package, and they will forget. Doing it server-side somewhere with some API is just a huge amount of maintenance incoming.

What will be the trigger for the updates then? We can maybe take care of this during `session.load()`. Once we get the drivers and results data back we can call a helper to update the constants. This way there is no extra API calls

**Comment by theOehrly** [comment]
> With the old API, every time a new driver took part in FP1 we had to manually input his abbreviation, team affiliation, and color. My understanding of the current API is we will still have to change the constants to map the drivers to teams. 

No, that is one of the big features of the new API. Drivers are automatically associated with teams. Only new teams require an update of the constants.

> > My idea was to implement this directly in FastF1. Updating the code base requires people to update the package, and they will forget. Doing it server-side somewhere with some API is just a huge amount of maintenance incoming.
> 
> What will be the trigger for the updates then? We can maybe take care of this during `session.load()`. Once we get the drivers and results data back we can call a helper to update the constants. This way there is no extra API calls

All functions from the new plotting API take ``session`` as a required argument. The driver-to-team mapping is automatically generated for each session based on the driver list from the livetiming API on the first call of a plotting function for this session. 
The driver list also provides the official team names and the official team color. Therefore, here we can automatically generate the constants for an unknown season. The only thing that will be missing is the FastF1 color map and the short names.

Doing any of this globally during ``session.load()`` would fail when users load different sessions for comparison.
